File: app/schemas/dasha.py
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DashaInterpretation(BaseModel):
    activated_houses: list[int] = Field(default_factory=list, alias="activatedHouses")
    house_text_ta: str = Field(default="", alias="houseTextTa")
    house_text_en: str = Field(default="", alias="houseTextEn")
    natural_domain_ta: str = Field(default="", alias="naturalDomainTa")
    natural_domain_en: str = Field(default="", alias="naturalDomainEn")
    relationship_to_maha_ta: str = Field(default="", alias="relationshipToMahaTa")
    relationship_to_maha_en: str = Field(default="", alias="relationshipToMahaEn")

    model_config = ConfigDict(populate_by_name=True)


class DashaPeriodWindow(BaseModel):
    lord: str
    start_date: date = Field(alias="startDate")
    end_date: date = Field(alias="endDate")
    interpretation: DashaInterpretation | None = Field(default=None)

    model_config = ConfigDict(populate_by_name=True)


class DashaOpeningWindow(BaseModel):
    lord: str
    balance_years_at_birth: float = Field(alias="balanceYearsAtBirth")

    model_config = ConfigDict(populate_by_name=True)


class DashaCurrentWindow(BaseModel):
    mahadasha: DashaPeriodWindow
    antardasha: DashaPeriodWindow
    pratyantardasha: DashaPeriodWindow


class DashaTimelineResponseData(BaseModel):
    chart_id: UUID = Field(alias="chartId")
    opening_dasha: DashaOpeningWindow = Field(alias="openingDasha")
    current: DashaCurrentWindow
    timeline: list[dict[str, object]] = Field(default_factory=list)

    model_config = ConfigDict(populate_by_name=True)


class ResponseMeta(BaseModel):
    calculation_version: str = Field(alias="calculationVersion")
    generated_at: datetime = Field(alias="generatedAt")

    model_config = ConfigDict(populate_by_name=True)


class DashaTimelineResponse(BaseModel):
    success: bool = True
    data: DashaTimelineResponseData
    meta: ResponseMeta

    model_config = ConfigDict(populate_by_name=True)

File: packages/shared/src/api/dasha.ts
import { getApiClient } from "./client";

export type DashaLevel = "maha" | "antar" | "pratyantar" | "sookshma" | "prana";

export interface DashaInterpretation {
  activatedHouses: number[];
  houseTextTa: string;
  houseTextEn: string;
  naturalDomainTa: string;
  naturalDomainEn: string;
  relationshipToMahaTa: string;
  relationshipToMahaEn: string;
}

export interface DashaTransitionNote {
  transitionDate: string;
  noteTa: string;
  noteEn: string;
}

export interface DashaPeriodWindow {
  lord: string;
  startDate: string;
  endDate: string;
  interpretation?: DashaInterpretation | null;
  transitionNote?: DashaTransitionNote | null;
  maturationStatus?: Record<string, unknown> | null;
}

export interface DashaTimelineItem extends DashaPeriodWindow {
  level: DashaLevel;
}

export interface DashaTimelineData {
  chartId: string;
  openingDasha: {
    lord: string;
    balanceYearsAtBirth: number;
  };
  current: {
    mahadasha: DashaPeriodWindow;
    antardasha: DashaPeriodWindow;
    pratyantardasha: DashaPeriodWindow;
  };
  timeline: DashaTimelineItem[];
}

export const dashaKeys = {
  timeline: (chartId: string, level: DashaLevel | readonly DashaLevel[] = "maha") =>
    ["dasha-timeline", chartId, Array.isArray(level) ? level.join(",") : level] as const,
};

/**
 * Dasha timeline. `level` accepts one level or a list — with a list, the
 * backend returns every requested level's rows concatenated in `timeline`
 * (each row carries its own `level` tag), so callers rendering maha + antar +
 * pratyantar views need a single request instead of three (DASH-04).
 * Backend: GET /api/v1/charts/{chart_id}/dasha (app/api/charts.py::get_dasha,
 * comma-separated `level` handled in app/services/dasha_service.py).
 */
export function getDashaTimeline(
  chartId: string,
  level: DashaLevel | readonly DashaLevel[] = "maha",
  asOf?: string,
): Promise<{ success: boolean; data: DashaTimelineData }> {
  return getApiClient().get(
    `/charts/${chartId}/dasha`,
    { level: Array.isArray(level) ? level.join(",") : (level as string), asOf },
  ) as Promise<{ success: boolean; data: DashaTimelineData }>;
}
